src/mlpack/core/optimizers/fw/frank_wolfe.hpp

```diff
@@ -0,0 +1,166 @@
+ * That is,
+ *
+ * \f[
+ * g(x):= \max_{s\in D} <x-s, \nabla f(x)> \quad \leq \epsilon,
+ * \f]
+ *
+ * we also know that \f$ g(x) \geq f(x) - f(x^*) \f$, where \f$ x^* \f$ is the optimal
+ * solution.
+ *
+ * The parameter \f$ \epsilon \f$ is specified by the tolerance parameter to the
+ * constructor.
+ *
+ * For FrankWolfe to work, LinearConstrSolverType and UpdateRuleType 
+ * template parameters are required.
+ * These classes must implement the following functions:
+ *
+ * LinearConstrSolverType:
+ *
+ *   void Optimize(const arma::mat& gradient,
+ *                 arma::mat& s);
+ *
+ * UpdateRuleType:
+ *
+ *   void Update(const arma::mat& old_coords,
+ *               const arma::mat& s,
+ *               arma::mat& new_coords,
+ *               const size_t num_iter);
+ *
+ * @tparam LinearConstrSolverType Solver for the linear constrained problem.
+ * @tparam UpdateRuleType Rule to update the solution in each iteration.
+ *
+ */
+template<
+    typename LinearConstrSolverType,
+    typename UpdateRuleType>
+class FrankWolfe
+{
+ public:
+  /**
+   * Construct the Frank-Wolfe optimizer with the given function and
+   * parameters. Notice that the constraint domain \f$ D \f$ is input
+   * at the initialization of linear_constr_solver, the function to be  
+   * optimized is stored in update_rule.
+   *
+   * @param linearConstrSolver Solver for linear constrained problem.
+   * @param updateRule Rule for updating solution in each iteration.
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *     limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   */
+  FrankWolfe(const LinearConstrSolverType linearConstrSolver,
+             const UpdateRuleType updateRule,
+             const size_t maxIterations = 100000,
+             const double tolerance = 1e-10);
+
+  /**
+   * Optimize the given function using FrankWolfe.  The given starting
+   * point will be modified to store the finishing point of the algorithm,
+   * the final objective value is returned.
+   *
+   * FunctionType template class must provide the following functions:
+   *
+   *   double Evaluate(const arma::mat& coordinates);
+   *   void Gradient(const arma::mat& coordinates,
+   *                 arma::mat& gradient);
+   *
+   * @param function Function to be optimized.
+   * @param iterate Input with starting point, and will be modified to save 
+   *                the output optimial solution coordinates.
+   * @return Objective value at the final solution.
+   */
+  template<typename FunctionType>
+  double Optimize(FunctionType& function, arma::mat& iterate);
+
+  //! Get the linear constrained solver.
+  LinearConstrSolverType LinearConstrSolver()
+      const { return linearConstrSolver; }
+  //! Modify the linear constrained solver.
+  LinearConstrSolverType& LinearConstrSolver() { return linearConstrSolver; }
+
+  //! Get the update rule.
+  UpdateRuleType UpdateRule() const { return updateRule; }
```

Comment:
The type here should be `const UpdateRuleType&`, otherwise this will cause a copy.  This applies to each of the other functions of this type, could you fix this please?